go/engine/merkle_audit.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2019 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+// MerkleAudit runs a merkle tree audit in the background once in a while.
+// It verifies the skips of a randomly chosen merkle tree root, making
+// sure that the server is not tampering with the merkle trees.
+
+package engine
+
+import (
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+)
+
+var (
+	errAuditOffline    = errors.New("Merkle audit failed to run due to the lack of connectivity.")
+	errAuditNoLastRoot = errors.New("Merkle audit failed to run due to not being able to get the last root.")
+)
+
+var MerkleAuditSettings = BackgroundTaskSettings{
+	Start:        5 * time.Minute,
+	StartStagger: 1 * time.Hour,
+	Interval:     6 * time.Hour,
+	Limit:        1 * time.Minute,
+}
+
+// MerkleAudit is an engine.
+type MerkleAudit struct {
+	libkb.Contextified
+	sync.Mutex
+
+	args *MerkleAuditArgs
+	task *BackgroundTask
+}
+
+type MerkleAuditArgs struct {
+	// Channels used for testing. Normally nil.
+	testingMetaCh     chan<- string
+	testingRoundResCh chan<- error
+}
+
+type merkleAuditState struct {
+	RetrySeqno *keybase1.Seqno `json:"retrySeqno"`
+}
+
+// NewMerkleAudit creates a new MerkleAudit engine.
+func NewMerkleAudit(g *libkb.GlobalContext, args *MerkleAuditArgs) *MerkleAudit {
+	task := NewBackgroundTask(g, &BackgroundTaskArgs{
+		Name:     "MerkleAudit",
+		F:        MerkleAuditRound,
+		Settings: MerkleAuditSettings,
+
+		testingMetaCh:     args.testingMetaCh,
+		testingRoundResCh: args.testingRoundResCh,
```

Comment:
Searching does not find these channels being used anywhere. Did I miss anything?

Reply:
These channels are used in the tests, I copied this approach from the PUK upgrader.

Reply:
ah ok, they are written to in `func (e *BackgroundTask) loop(m libkb.MetaContext) error`. sorry.